// @polymarket/clob-client/dist/http-helpers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPriceHistoryFilterParamsToUrl = exports.addOrdersScoringParamsToUrl = exports.addOrderScoringParamsToUrl = exports.addBalanceAllowanceParamsToUrl = exports.addTradeNotificationParamsToUrl = exports.addTradeParamsToUrl = exports.addOpenOrderParamsToUrl = exports.addFilterParamsToUrl = exports.buildQueryParams = exports.del = exports.get = exports.post = exports.request = exports.PUT = exports.DELETE = exports.POST = exports.GET = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const browser_or_node_1 = require("browser-or-node");
exports.GET = "GET";
exports.POST = "POST";
exports.DELETE = "DELETE";
exports.PUT = "PUT";
const overloadHeaders = (method, headers) => {
    if (browser_or_node_1.isBrowser) {
        return;
    }
    if (!headers || typeof headers === undefined) {
        headers = {};
    }
    if (headers) {
        headers["User-Agent"] = `@polymarket/clob-client`;
        headers["Accept"] = "*/*";
        headers["Connection"] = "keep-alive";
        headers["Content-Type"] = "application/json";
        if (method === exports.GET) {
            headers["Accept-Encoding"] = "gzip";
        }
    }
};
const request = (endpoint, method, headers, data, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d;
    try {
        overloadHeaders(method, headers);
        const response = yield (0, axios_1.default)({ method, url: endpoint, headers, data, params });
        return response;
    }
    catch (err) {
        if (axios_1.default.isAxiosError(err)) {
            if (err.response) {
                console.error("request error", {
                    status: (_a = err.response) === null || _a === void 0 ? void 0 : _a.status,
                    statusText: (_b = err.response) === null || _b === void 0 ? void 0 : _b.statusText,
                    data: (_c = err.response) === null || _c === void 0 ? void 0 : _c.data,
                });
                return (_d = err.response) === null || _d === void 0 ? void 0 : _d.data;
            }
            else {
                return { error: "connection error" };
            }
        }
        return { error: err };
    }
});
exports.request = request;
const post = (endpoint, headers, data, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const resp = yield (0, exports.request)(endpoint, exports.POST, headers, data, params);
    return "error" in resp ? resp : resp.data;
});
exports.post = post;
const get = (endpoint, headers, data, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const resp = yield (0, exports.request)(endpoint, exports.GET, headers, data, params);
    return "error" in resp ? resp : resp.data;
});
exports.get = get;
const del = (endpoint, headers, data, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const resp = yield (0, exports.request)(endpoint, exports.DELETE, headers, data, params);
    return "error" in resp ? resp : resp.data;
});
exports.del = del;
const buildQueryParams = (url, param, value) => {
    let urlWithParams = url;
    const last = url.charAt(url.length - 1);
    // Check the last char in the url string
    // if ?, append the param directly: api.com?param=value
    if (last === "?") {
        urlWithParams = `${urlWithParams}${param}=${value}`;
    }
    else {
        // else append "&" then the param: api.com?param1=value1&param2=value2
        urlWithParams = `${urlWithParams}&${param}=${value}`;
    }
    return urlWithParams;
};
exports.buildQueryParams = buildQueryParams;
const addFilterParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.owner !== undefined) {
            url = (0, exports.buildQueryParams)(url, "owner", params.owner);
        }
        if (params.max !== undefined) {
            url = (0, exports.buildQueryParams)(url, "max", `${params.max}`);
        }
        if (params.market !== undefined) {
            url = (0, exports.buildQueryParams)(url, "market", params.market);
        }
        if (params.side !== undefined) {
            url = (0, exports.buildQueryParams)(url, "side", params.side);
        }
        if (params.startTs !== undefined) {
            url = (0, exports.buildQueryParams)(url, "startTs", `${params.startTs}`);
        }
        if (params.endTs !== undefined) {
            url = (0, exports.buildQueryParams)(url, "endTs", `${params.endTs}`);
        }
        if (params.minValue !== undefined) {
            url = (0, exports.buildQueryParams)(url, "minValue", `${params.minValue}`);
        }
        if (params.fidelity !== undefined) {
            url = (0, exports.buildQueryParams)(url, "fidelity", `${params.fidelity}`);
        }
    }
    return url;
};
exports.addFilterParamsToUrl = addFilterParamsToUrl;
const addOpenOrderParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.market !== undefined) {
            url = (0, exports.buildQueryParams)(url, "market", params.market);
        }
        if (params.asset_id !== undefined) {
            url = (0, exports.buildQueryParams)(url, "asset_id", params.asset_id);
        }
        if (params.owner !== undefined) {
            url = (0, exports.buildQueryParams)(url, "owner", params.owner);
        }
        if (params.id !== undefined) {
            url = (0, exports.buildQueryParams)(url, "id", params.id);
        }
    }
    return url;
};
exports.addOpenOrderParamsToUrl = addOpenOrderParamsToUrl;
const addTradeParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.owner !== undefined) {
            url = (0, exports.buildQueryParams)(url, "owner", params.owner);
        }
        if (params.market !== undefined) {
            url = (0, exports.buildQueryParams)(url, "market", params.market);
        }
        if (params.asset_id !== undefined) {
            url = (0, exports.buildQueryParams)(url, "asset_id", params.asset_id);
        }
        if (params.maker !== undefined) {
            url = (0, exports.buildQueryParams)(url, "maker", `${params.maker}`);
        }
        if (params.taker !== undefined) {
            url = (0, exports.buildQueryParams)(url, "taker", `${params.taker}`);
        }
        if (params.id !== undefined) {
            url = (0, exports.buildQueryParams)(url, "id", `${params.id}`);
        }
        if (params.limit !== undefined) {
            url = (0, exports.buildQueryParams)(url, "limit", `${params.limit}`);
        }
        if (params.before !== undefined) {
            url = (0, exports.buildQueryParams)(url, "before", `${params.before}`);
        }
        if (params.after !== undefined) {
            url = (0, exports.buildQueryParams)(url, "after", `${params.after}`);
        }
    }
    return url;
};
exports.addTradeParamsToUrl = addTradeParamsToUrl;
const addTradeNotificationParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.index !== undefined) {
            url = (0, exports.buildQueryParams)(url, "index", params.index.toString());
        }
    }
    return url;
};
exports.addTradeNotificationParamsToUrl = addTradeNotificationParamsToUrl;
const addBalanceAllowanceParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.asset_type !== undefined) {
            url = (0, exports.buildQueryParams)(url, "asset_type", params.asset_type.toString());
        }
        if (params.token_id && params.token_id !== undefined) {
            url = (0, exports.buildQueryParams)(url, "token_id", params.token_id.toString());
        }
    }
    return url;
};
exports.addBalanceAllowanceParamsToUrl = addBalanceAllowanceParamsToUrl;
const addOrderScoringParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.orderId !== undefined) {
            url = (0, exports.buildQueryParams)(url, "order_id", params.orderId.toString());
        }
    }
    return url;
};
exports.addOrderScoringParamsToUrl = addOrderScoringParamsToUrl;
const addOrdersScoringParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.orderIds !== undefined) {
            url = (0, exports.buildQueryParams)(url, "order_ids", params.orderIds.join(","));
        }
    }
    return url;
};
exports.addOrdersScoringParamsToUrl = addOrdersScoringParamsToUrl;
const addPriceHistoryFilterParamsToUrl = (baseUrl, params) => {
    let url = baseUrl;
    if (params !== undefined) {
        url = `${url}?`;
        if (params.market !== undefined) {
            url = (0, exports.buildQueryParams)(url, "market", params.market);
        }
        if (params.startTs !== undefined) {
            url = (0, exports.buildQueryParams)(url, "startTs", `${params.startTs}`);
        }
        if (params.endTs !== undefined) {
            url = (0, exports.buildQueryParams)(url, "endTs", `${params.endTs}`);
        }
        if (params.fidelity !== undefined) {
            url = (0, exports.buildQueryParams)(url, "fidelity", `${params.fidelity}`);
        }
        if (params.interval !== undefined) {
            url = (0, exports.buildQueryParams)(url, "interval", `${params.interval}`);
        }
    }
    return url;
};
exports.addPriceHistoryFilterParamsToUrl = addPriceHistoryFilterParamsToUrl;
