// @polymarket/clob-client/dist/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GET_MARKET_TRADES_EVENTS = exports.GET_BALANCE_ALLOWANCE = exports.DROP_TRADE_NOTIFICATIONS = exports.GET_TRADE_NOTIFICATIONS = exports.GET_PRICES_HISTORY = exports.ARE_ORDERS_SCORING = exports.IS_ORDER_SCORING = exports.GET_TRADES = exports.GET_OPEN_ORDERS = exports.GET_LARGE_ORDERS = exports.CANCEL_MARKET_ORDERS = exports.CANCEL_ALL = exports.GET_ORDER = exports.CANCEL_ORDERS = exports.CANCEL_ORDER = exports.POST_ORDER = exports.GET_TICK_SIZE = exports.GET_LAST_TRADE_PRICE = exports.GET_PRICE = exports.GET_MIDPOINT = exports.GET_ORDER_BOOK = exports.GET_MARKET = exports.GET_MARKETS = exports.GET_SIMPLIFIED_MARKETS = exports.GET_SAMPLING_MARKETS = exports.GET_SAMPLING_SIMPLIFIED_MARKETS = exports.DERIVE_API_KEY = exports.DELETE_API_KEY = exports.GET_API_KEYS = exports.CREATE_API_KEY = exports.TIME = void 0;
// Server Time
exports.TIME = "/time";
// API Key endpoints
exports.CREATE_API_KEY = "/create-api-key";
exports.GET_API_KEYS = "/get-api-keys";
exports.DELETE_API_KEY = "/delete-api-key";
exports.DERIVE_API_KEY = "/derive-api-key";
// Markets
exports.GET_SAMPLING_SIMPLIFIED_MARKETS = "/sampling-simplified-markets";
exports.GET_SAMPLING_MARKETS = "/sampling-markets";
exports.GET_SIMPLIFIED_MARKETS = "/simplified-markets";
exports.GET_MARKETS = "/markets";
exports.GET_MARKET = "/markets/";
exports.GET_ORDER_BOOK = "/book";
exports.GET_MIDPOINT = "/midpoint";
exports.GET_PRICE = "/price";
exports.GET_LAST_TRADE_PRICE = "/last-trade-price";
exports.GET_TICK_SIZE = "/tick-size";
// Order endpoints
exports.POST_ORDER = "/order";
exports.CANCEL_ORDER = "/order";
exports.CANCEL_ORDERS = "/orders";
exports.GET_ORDER = "/order/";
exports.CANCEL_ALL = "/cancel-all";
exports.CANCEL_MARKET_ORDERS = "/cancel-market-orders";
exports.GET_LARGE_ORDERS = "/large-orders";
exports.GET_OPEN_ORDERS = "/orders";
exports.GET_TRADES = "/trades";
exports.IS_ORDER_SCORING = "/order-scoring";
exports.ARE_ORDERS_SCORING = "/orders-scoring";
// Price history
exports.GET_PRICES_HISTORY = "/prices-history";
// Notifications
exports.GET_TRADE_NOTIFICATIONS = "/trade-notifications";
exports.DROP_TRADE_NOTIFICATIONS = "/drop-trade-notifications";
// Balance
exports.GET_BALANCE_ALLOWANCE = "/balance-allowance";
// Live activity
exports.GET_MARKET_TRADES_EVENTS = "/live-activity/events/";
